hooks/useDeposit.tsx

```diff
@@ -1,72 +1,117 @@
-
-  const { config } = usePrepareContractWrite({
+  // Simulate the contract call first to validate it
+  const { data: simulateData, isError: isSimulateError } = useSimulateContract({
     address: contractAddress || '',
     abi: WethAbi,
     functionName: 'deposit',
-    enabled: Boolean(debouncedValue),
     account: address,
-    value: BigInt(parseEther(debouncedValue.toString() || '0')),
-    onSuccess(data: any) {
-      return data;
-    },
-    onError(error: any) {
-      return error;
-    },
+    value: simulateValue,
   });
 
-  const { write: writeDeposit, data: dataDeposit } = useContractWrite({
-    ...config,
-    request: config.request,
-    onSuccess() {
-      toast.success({
-        title: 'Pending Transaction...',
-        isClosable: true,
-      });
-    },
-    onError() {
-      toast.error({
-        title: 'Error... transaction reverted...',
-        isClosable: true,
-      });
+  // Use the writeContract hook to execute the transaction
+  const {
+    writeContract: writeDeposit,
+    data: dataDeposit,
+    isPending: isWritePending,
+    isError: isWriteError,
+  } = useWriteContract({
+    mutation: {
+      onSuccess() {
+        toast.success({
+          title: 'Pending Transaction...',
+          isClosable: true,
+        });
+      },
+      onError() {
+        toast.error({
+          title: 'Error... transaction reverted...',
+          isClosable: true,
+        });
+      },
     },
   });
 
-  const { status: statusDeposit } = useWaitForTransaction({
-    hash: dataDeposit?.hash,
-    onSuccess: () => {
-      toast.success({
-        title: `Success! Wrapped ${chain?.nativeCurrency?.symbol || 'ETH'}`,
-        isClosable: true,
-      });
-    },
+  // Add a ref to track if we've already shown the toast for this transaction
+  const hasShownToastRef = useRef<{ [txHash: string]: boolean }>({});
+
+  // Wait for the transaction receipt
+  const {
+    data: receiptData,
+    isLoading: isConfirming,
+    isSuccess: isConfirmed,
+    status: statusDeposit
+  } = useWaitForTransactionReceipt({
+    hash: dataDeposit,
+    query: {
+      refetchOnWindowFocus: false,
+    }
   });
 
+  // Use useEffect to handle the toast only once per transaction
+  useEffect(() => {
+    if (isConfirmed && receiptData && dataDeposit) {
+      // Check if we've already shown a toast for this transaction
+      if (!hasShownToastRef.current[dataDeposit]) {
+        toast.success({
+          title: `Success! Wrapped ${chain?.nativeCurrency?.symbol || 'ETH'}`,
+          isClosable: true,
+        });
+
+        // Mark this transaction as having shown a toast
+        hasShownToastRef.current[dataDeposit] = true;
+      }
+    }
+  }, [isConfirmed, receiptData, dataDeposit, chain?.nativeCurrency?.symbol, toast]);
+
+  // Function to execute the deposit
+  const executeDeposit = () => {
+    if (simulateData?.request) {
+      writeDeposit(simulateData.request);
```

Comment:
if you use the async version, it'll help clean up this hook a decent amount I think. Use the viem `waitForTransaction` after waiting for the hash here. Gets you out of this annoying useEffect pattern.

Technically you don't need to simulate these anymore, if we're making proper frontend checks of the input values. This is a holdover from the v0 version of wagmi.